### ISSUE TITLE
Underscores in numbers in script blocks break minification

### ISSUE THREAD

**Issue by curio77** [tdewolff/minify#466]
Using an underscore in a JS integer as a numeric separator:

```go
package main

import (
        "os"
        "regexp"
        "strings"

        "github.com/tdewolff/minify/v2"
        "github.com/tdewolff/minify/v2/html"
        "github.com/tdewolff/minify/v2/js"
)

func main() {
        input := `<script>const i = 1_000</script>`
        m := minify.New()
        m.AddFunc("text/html", html.Minify)
        m.AddFuncRegexp(regexp.MustCompile("^(application|text)/(x-)?(java|ecma)script$"), js.Minify)

        if err := m.Minify("text/html", os.Stdout, strings.NewReader(input)); err != nil {
                panic(err)
        }
}
```

results in this error:

```
<script>panic: unexpected identifier after number on line 1 and column 24
    1: const i = 1_000
```

Not sure if this is expected or whether minify should cope with that.

What's worse, however, is that if this occurs with minification via the middleware, no error is reported in any way and output truncated after the preceding `<script>` tag, i.e., this:

```go
package main

import (
        "fmt"
        "net/http"
        "net/http/httptest"
        "regexp"

        "github.com/tdewolff/minify/v2"
        "github.com/tdewolff/minify/v2/html"
        "github.com/tdewolff/minify/v2/js"
)

func main() {
        m := minify.New()
        m.AddFunc("text/html", html.Minify)
        m.AddFuncRegexp(regexp.MustCompile("^(application|text)/(x-)?(java|ecma)script$"), js.Minify)

        input := `<script>const i = 1_000</script>`
        req := httptest.NewRequest(http.MethodGet, "/", nil)
        rec := httptest.NewRecorder()
        m.Middleware(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
                w.Header().Set("Content-Type", "text/html")
                _, err := w.Write([]byte(input))
                if err != nil {
                        panic(err)
                }
        })).ServeHTTP(rec, req)
        fmt.Println(rec.Body.String())
}
```

yields just `<script>`.

**Comment by tdewolff** [comment]
Sorry for the delay. Underscores in numbers is a relatively new feature that I'll work on to support. The second point, the error should indeed be passed to the user, I will look into that.

**Comment by tdewolff** [comment]
As to your second problem with error propagation. The problem is that when you call `Write`, it immediately returns while minifying concurrently. At that point no error has been found, which is why the returned error is nil. The most reliable way is to call `err = w.Close()` at the end of the `HandlerFunc` and check the error:

```go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"regexp"

	"github.com/tdewolff/minify/v2"
	"github.com/tdewolff/minify/v2/html"
	"github.com/tdewolff/minify/v2/js"
)

type ResponseWriterCloser interface {
	http.ResponseWriter
	Close() error
}

func main() {
	m := minify.New()
	m.AddFunc("text/html", html.Minify)
	m.AddFuncRegexp(regexp.MustCompile("^(application|text)/(x-)?(java|ecma)script$"), js.Minify)

	input := `<script>const i = 1_000_</script>`
	req := httptest.NewRequest(http.MethodGet, "/", nil)
	rec := httptest.NewRecorder()
	m.Middleware(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/html")
		_, err := w.Write([]byte(input))
		if err != nil {
			panic(err)
		}

		err = w.(ResponseWriterCloser).Close()
		if err != nil {
			panic(err)
		}
	})).ServeHTTP(rec, req)
	fmt.Println(rec.Body.String())
}
```

I agree this is not the most ergonomic...I'd be very happy to hear a better solution!

**Comment by curio77** [comment]
Thank you very much for looking into this! :)

About closing the `ResponseWriter`... I assume that the `Close()` method is normally called in the Minify middleware after the “next” handler returns? To implement this generally (in a fashion that allows reacting to the error), I guess it would be best to implement a wrapper for or dedicated “next” middleware that essentially calls `Close()` after the (actual) next handler?  In this case, might it be a cleaner approach to have a flag in creating the Minify middleware that disables concurrent/background minification?

On the other hand, it might be better to have Minify render to a buffer anyway in a custom middleware, giving the option to render an error response instead to the `ResponseWriter` on errors, if one wants to be able to cope with that.  Maybe only in dev mode and not generally.  But then I don't have minification enabled in dev mode any… ;-)

In any case, I might be worth documenting the caveat (that using the middleware just like any other will yield incomplete output on errors, and no error on write).

**Comment by tdewolff** [comment]
Good points, made me think. What about added a second argument to `Middleware`? As in `Middleware(next http.Handler, handleError func(error))` which lets you handle it yourself.

**Comment by curio77** [comment]
Yes, or rather make it a separate method to avoid requiring existing users to change their code?

Of course, that error handler is asynchronous in that it's not in the request handling pipeline, but it enables at the very least logging the error and thus noticing there's an issue somewhere.  For that, it would also be desirable to cross-reference the request whose handling triggered the error, so the error message should include, say, the request URL or path (for want of a more content-related reference, which the middleware cannot provide as it doesn't know the source of the document/template rendition/whatever it got passed).  Does that make sense?

**Comment by tdewolff** [comment]
I've added an example of using the current `Middleware` with

```go
if err := w.(io.Closer).Close(); err != nil {
	panic(err)
}
```

as well as the `MiddlewareWithError`  function, which accepts as second parameter the signature `func(w http.ResponseWriter, r *http.Request, err error)`. What do you think?

**Comment by curio77** [comment]
Looks good!  The `http.ResponseWriter` argument is probably not of much use (if the Minify middleware has already started writing a response to it), but maybe there *are* cases where it may come in handy.

**Comment by tdewolff** [comment]
This has been fixed and released in v.2.11.0!
